src/main/scala/mesosphere/marathon/core/election/impl/CuratorElectionService.scala

```diff
@@ -125,6 +126,25 @@ class CuratorElectionService(
           CurrentRuntime.asyncExit()(scala.concurrent.ExecutionContext.global)
           false
         }
+      }).zookeeperFactory(new ZookeeperFactory {
+        val delegate = new DefaultZookeeperFactory
+        override def newZooKeeper(
+          connectString: String,
+          sessionTimeout: Int,
+          watcher: Watcher,
+          canBeReadOnly: Boolean): ZooKeeper =
+
+          try {
+            delegate.newZooKeeper(connectString, sessionTimeout, watcher, canBeReadOnly)
+          } catch {
+            // We may see exceptions here, for example, when none of the ZK hosts are resolvable at Marathon startup;
+            // in this case Curator may fail with some background exception and Marathon never offers leadership.
+            case NonFatal(e) =>
+              // TODO(jdef) should we really be committing suicide here? what about retries instead?
+              log.error("ZooKeeper initialization failed - Committing suicide")
+              CurrentRuntime.asyncExit()(scala.concurrent.ExecutionContext.global)
+              null
```

Comment:
My initial comment was re: the `null`.

However, since you've brought up DNS... I'm told that on dc/os we use special hostnames that should resolve to virtual IPs. Precisely because we can't trust Java DNS caching. I suspect we're still seeing this in dc/os at bootstrapping time because the service providing that function isn't available yet.

However, we should also degrade gracefully in non-dc/os envs. Anyone running Marathon in production should be running it with a service-runner (systemd, runit, whatever) so that if Marathon dies it will respawned shortly.

The solution as implemented is, IMO, a decent compromise because it (a) makes no assumptions about DNS; (b) makes no assumption about dc/os vs mesos, and; (c) relies on (safe) assumptions re: production-ready deployment.

Alternatives? We could build in our own retry mechanism. Or provide our own host-lookup provider. I'm not convinced that such band-aids are any better. If the community clamors for additional flexibility we could create a plugin interface for providing custom ZooKeeperFactory implementations and just let people bring their own if they hate our default behavior.